### ISSUE TITLE
Build fails on OS X 10.11.5 (commit 64380109)

### ISSUE THREAD

**Issue by hatarist** [yandex/ClickHouse#232]
I followed the [latest build_osx.md guide](https://github.com/yandex/ClickHouse/blob/64380109e956e462b6a98263e8bd59fd8ebb52cc/doc/build_osx.md).

```
❯ brew list --versions | egrep '(cmake|gcc|boost|icu4c|mysql|openssl|unixodbc|glib|libtool|gettext|libiconv|zlib|readline)' | sort
boost 1.62.0
cmake 3.7.1
gcc 6.2.0
gettext 0.19.8.1
glib 2.50.2
icu4c 58.1
libiconv 1.14
libtool 2.4.6_1
mysql 5.7.16
openssl 1.0.2j
openssl@1.1 1.1.0c
readline 7.0.1
unixodbc 2.3.4
zlib 1.2.8
```

`cmake ..` crashes on iconv. Full log: http://pastebin.ca/3744809
```
Creating headers for re2_st library.
-- Building: tcmalloc_minimal_internal
-- Link libtcmalloc_minimal
-- Using line editing libraries: /usr/local/opt/readline/lib/libreadline.a;/usr/lib/libtermcap.dylib
-- C_FLAGS:  =-fno-omit-frame-pointer -Wall -Werror -msse4 -mpopcnt  -D_GLIBCXX_USE_CXX11_ABI=0
-- CXX_FLAGS:=-std=gnu++1y -Dexp10=__exp10 -fno-omit-frame-pointer -Wall -Werror -Wnon-virtual-dtor -Wold-style-cast -msse4 -mpopcnt  -D_GLIBCXX_USE_CXX11_ABI=0
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
ICONV_LIB
    linked by target "common" in directory /Users/igor/Projects/clickhouse/libs/libcommon

-- Configuring incomplete, errors occurred!
See also "/Users/igor/Projects/clickhouse/build/CMakeFiles/CMakeOutput.log".
```

CMakeOutput.log: http://pastebin.ca/3744811

libiconv info:
```
❯ brew info libiconv
homebrew/dupes/libiconv: stable 1.14 (bottled) [keg-only]
Conversion library
https://www.gnu.org/software/libiconv/
/usr/local/Cellar/libiconv/1.14 (28 files, 1.4M)
  Poured from bottle on 2016-12-07 at 11:59:23
From: https://github.com/Homebrew/homebrew-dupes/blob/master/libiconv.rb
<...>

❯ ls -l /usr/local/opt/libiconv/lib
total 2120
-r--r--r--  1 igor  admin    13184 Dec  7 11:59 libcharset.1.dylib
-r--r--r--  1 igor  admin     3720 Oct 30  2015 libcharset.a
lrwxr-xr-x  1 igor  admin       18 Oct 30  2015 libcharset.dylib -> libcharset.1.dylib
-r--r--r--  1 igor  admin  1056200 Dec  7 11:59 libiconv.2.dylib
lrwxr-xr-x  1 igor  admin       16 Oct 30  2015 libiconv.dylib -> libiconv.2.dylib
```

Removal of the homebrew's iconv doesn't seem to change anything.

I also tried to replace `libiconv.a` with `libiconv.dylib` in the `cmake/find_glib.cmake` file:
```
-		find_library (ICONV_LIB libiconv.a HINTS ${ICONV_HINTS})
+		find_library (ICONV_LIB libiconv.dylib HINTS ${ICONV_HINTS})
```

This way, `cmake ..` ran without errors. But ti failed on the `make` (more lines at http://pastebin.ca/3744814):
```
Undefined symbols for architecture x86_64:
  "_iconv", referenced from:
      _g_iconv in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      _g_convert_with_iconv in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      _g_convert_with_fallback in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      __nl_find_msg in libintl.a(dcigettext.o)
     (maybe you meant: _g_iconv_open, _g_iconv , _g_iconv_close , _g_convert_with_iconv )
  "_iconv_close", referenced from:
      _g_iconv_close in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      _g_convert in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      _g_convert_with_fallback in libglib-2.0.a(libglib_2_0_la-gconvert.o)
     (maybe you meant: _g_iconv_close)
  "_iconv_open", referenced from:
      _try_conversion in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      __nl_find_msg in libintl.a(dcigettext.o)
     (maybe you meant: _g_iconv_open)
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [libs/libcommon/src/tests/date_lut2] Error 1
make[1]: *** [libs/libcommon/src/tests/CMakeFiles/date_lut2.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
I'm not sure what to do here, sorry, since I have very little experience with cmake/makefile :(

**Comment by hatarist** [comment]
I also tried to build it with the manually compiled libiconv-1.14.
I had errors like that:
```
[ 81%] Linking CXX executable date_lut2
Undefined symbols for architecture x86_64:
  "_iconv", referenced from:
      _g_iconv in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      _g_convert_with_iconv in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      _g_convert_with_fallback in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      __nl_find_msg in libintl.a(dcigettext.o)
     (maybe you meant: _g_iconv_open, _g_convert_with_iconv , _g_iconv , _g_iconv_close )
  "_iconv_close", referenced from:
      _g_iconv_close in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      _g_convert in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      _g_convert_with_fallback in libglib-2.0.a(libglib_2_0_la-gconvert.o)
     (maybe you meant: _g_iconv_close)
  "_iconv_open", referenced from:
      _try_conversion in libglib-2.0.a(libglib_2_0_la-gconvert.o)
      __nl_find_msg in libintl.a(dcigettext.o)
     (maybe you meant: _g_iconv_open)
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [libs/libcommon/src/tests/date_lut2] Error 1
make[1]: *** [libs/libcommon/src/tests/CMakeFiles/date_lut2.dir/all] Error 2
make: *** [all] Error 2
```

I tried setting and unsetting `CC=gcc-6` `CXX=g++-6` (also, `CPP="g++-6 -E"`) and configured with`./configure --enable-static` (with `make` / `make install` afterwards and `make clean` in between)

**Comment by hatarist** [comment]
BTW, I was able to successfully compile it using `USE_STATIC_LIBRARIES=0 cmake ..` but, obviously, it crashes on start after doing `brew uninstall libiconv`. But it is able to actually find libiconv that way.

**Comment by hatarist** [comment]
`USE_STATIC_LIBRARIES=0` is now (690ab98b9) broken, too. :(

```
[ 89%] Linking CXX executable glue_streams
Undefined symbols for architecture x86_64:
  "_iconv", referenced from:
      DB::FunctionConvertCharset::convert(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul> const&, DB::PODArray<unsigned long, 4096ul, Allocator<false>, 15ul> const&, DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul>&, DB::PODArray<unsigned long, 4096ul, Allocator<false>, 15ul>&) in libdbms.a(FunctionsCharset.cpp.o)
  "_iconv_close", referenced from:
      std::_Rb_tree<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const, std::unique_ptr<DB::ObjectPool<DB::FunctionConvertCharset::IConv, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::Holder, std::default_delete<DB::ObjectPool<DB::FunctionConvertCharset::IConv, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::Holder> > >, std::_Select1st<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const, std::unique_ptr<DB::ObjectPool<DB::FunctionConvertCharset::IConv, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::Holder, std::default_delete<DB::ObjectPool<DB::FunctionConvertCharset::IConv, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::Holder> > > >, std::less<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::allocator<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const, std::unique_ptr<DB::ObjectPool<DB::FunctionConvertCharset::IConv, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::Holder, std::default_delete<DB::ObjectPool<DB::FunctionConvertCharset::IConv, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::Holder> > > > >::_M_erase(std::_Rb_tree_node<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const, std::unique_ptr<DB::ObjectPool<DB::FunctionConvertCharset::IConv, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::Holder, std::default_delete<DB::ObjectPool<DB::FunctionConvertCharset::IConv, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::Holder> > > >*) in libdbms.a(FunctionsCharset.cpp.o)
      DB::ObjectPool<DB::FunctionConvertCharset::IConv, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::~ObjectPool() in libdbms.a(FunctionsCharset.cpp.o)
      std::deque<std::unique_ptr<DB::FunctionConvertCharset::IConv, std::default_delete<DB::FunctionConvertCharset::IConv> >, std::allocator<std::unique_ptr<DB::FunctionConvertCharset::IConv, std::default_delete<DB::FunctionConvertCharset::IConv> > > >::~deque() in libdbms.a(FunctionsCharset.cpp.o)
      DB::FunctionConvertCharset::convert(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul> const&, DB::PODArray<unsigned long, 4096ul, Allocator<false>, 15ul> const&, DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul>&, DB::PODArray<unsigned long, 4096ul, Allocator<false>, 15ul>&) in libdbms.a(FunctionsCharset.cpp.o)
  "_iconv_open", referenced from:
      DB::FunctionConvertCharset::IConv::IConv(std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const&) in libdbms.a(FunctionsCharset.cpp.o)
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [dbms/src/DataStreams/tests/glue_streams] Error 1
make[1]: *** [dbms/src/DataStreams/tests/CMakeFiles/glue_streams.dir/all] Error 2
make: *** [all] Error 2
```

I made sure to put the include/library paths:
```
export CFLAGS="-I/usr/local/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl/include -I/usr/local/opt/libiconv/include"
export C_FLAGS="-I/usr/local/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl/include -I/usr/local/opt/libiconv/include"

export CPPFLAGS="-I/usr/local/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl/include -I/usr/local/opt/libiconv/include"
export CPP_FLAGS="-I/usr/local/include -I/usr/local/opt/readline/include -I/usr/local/opt/openssl/include -I/usr/local/opt/libiconv/include"

export LDFLAGS="-L/usr/local/lib -L/usr/local/opt/readline/lib -L/usr/local/opt/openssl/lib -L/usr/local/opt/libiconv/lib"
export LD_FLAGS="-L/usr/local/lib -L/usr/local/opt/readline/lib -L/usr/local/opt/openssl/lib -L/usr/local/opt/libiconv/lib"
```
but it didn't help any :(

**Comment by hatarist** [comment]
Seems to be fixed after f63a3ab1. Thanks!
